scripts/semantic_version.sh

```diff
@@ -14,61 +14,33 @@
-#
-# The script looks in a few places for the version, in this order:
-#  - The TRAVIS_TAG environment variable trumps all: if present, the version contained here will be
-#    used. It may be a "release" build, e.g. the semantic version for the release tagged
-#    "linux-v1.0.0" is 1.0.0, or a pre-release build, e.g. the semantic version for the release
-#    tagged "daily-2018-12-01" is "0.0.0-daily-2018-12-01".
-#  - As an alternative to TRAVIS_TAG, tags associated with the current git will be searched. This is
-#    primarily intended for Windows release builds, which cannot be performed on Travis. For this to
-#    work, you must specify the -p switch which helps disambiguate commits with multiple release
-#    tags.
-#  - If the previous two searches fail, the short-form SHA of the current commit will be appended to
-#    the version 0.0.0, e.g. 0.0.0+a78f88.
-
-PLATFORM=
+# Outputs a semantic version for the app, intended for the given platform and environment.
 
 function usage () {
-  echo "$0 [-p platform] [-h]"
-  echo "  -p: platform"
-  echo "  -h: this help message"
-  echo
-  echo "Examples:"
-  echo "  $0 -p linux"
+  echo "$0 version platform environment" 1>&2
+  echo "  version: the new version of the application" 1>&2
+  echo "  platform: (android, ios, osx, browser, windows, linux, or dev)" 1>&2
+  echo "  environment: the environment you're deploying to (development, prerelease, production)" 1>&2
+  echo 1>&2
+  echo "Examples:" 1>&2
+  echo "  $0 1.2.2 android development" 1>&2
   exit 1
 }
 
-while getopts p:h? opt; do
-  case $opt in
-    p) PLATFORM=$OPTARG ;;
-    h) usage ;;
-    *) usage ;;
-  esac
-done
-shift $((OPTIND-1))
+SEMVER_VERSION=$1
+PLATFORM=$2
+ENVIRONMENT=$3
 
-# Search for a tag in the Travis environment and the current commit's tags.
-if ! TAG=$($(dirname $0)/get_tag.sh $PLATFORM); then
-  echo "Cannot compute semantic version" >&2
-  exit 2
-fi
+[[ ! $SEMVER_VERSION || ! $PLATFORM || ! $ENVIRONMENT ]] && usage
+[[ "${SEMVER_VERSION}" =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]] || exit 2
 
-# Search for a semantic version in the tag, e.g. windows-v1.0.1 -> 1.0.1.
-if echo "$TAG" | grep -q '.*-v'; then
-  SEMVER=$(echo "$TAG" | sed 's/[^-]*-v//')
-fi
+SEMVER_METADATA="+${PLATFORM}"
+SEMVER_PRERELEASE=
 
-if [[ -n "$SEMVER" ]]; then
-  # Production build.
-  echo "$SEMVER"
-elif [[ -n "$TAG" ]]; then
-  # Pre-release, e.g. daily build:
-  #   https://semver.org/#spec-item-9
-  echo "0.0.0-$TAG"
-else
-  # Local, ad-hoc run or build: just append the SHA of the current commit:
-  #   https://semver.org/#spec-item-10
-  echo "0.0.0+$(git log --pretty=format:'%h' -n 1)"
+if [[ "${ENVIRONMENT}" == "development" ]]; then
+  SEMVER_METADATA="+${PLATFORM}.$(git log --pretty=format:'%h' -n 1)"
+elif [[ "${ENVIRONMENT}" == "prerelease" ]]; then
+  SEMVER_PRERELEASE="-rc$(git tag -l "${SEMVER_VERSION}-rc*${SEMVER_METADATA}" | wc -l | xargs)"
 fi
+
+echo "${SEMVER_VERSION}${SEMVER_PRERELEASE}${SEMVER_METADATA}"
```

Comment:
I'm attempting to keep valid semver notation in the tags - https://semver.org/ - let me know what you think.